### PR TITLE
Add ConvertTo-Clixml and ConvertFrom-Clixml cmdlet

### LINKED ISSUE #3898: Export-CliXml shouldn't require writing to a file
The `Export-CliXml` command should offer a parameter set that emits a string, instead of writing the result to a file. Right now, it only supports outputting CLIXML to a file via the `-Path` or `-LiteralPath` parameters.

Cheers,
[Trevor Sullivan](https://trevorsullivan.net)

### PR DESCRIPTION
**Pull request by Shriram0908**
# PR Summary

```ConvertTo-Clixml``` and ```ConvertFrom-CliXml``` to work with clixml objects. Fixes [#3898](https://github.com/PowerShell/PowerShell/issues/3898)

## PR Context

The new cmdlet can be used to work with clixml without needed to save it file.
The cmdlets behave similar to ```ConvertTo-Csv``` and ```ConvertFrom-Csv```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: [#6093](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/6093) 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by Shriram0908** [comment]
Hi @iSazonov, Could you review the PR and provide your feedback.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:198]:
```suggestion
            if (!ShouldProcess(Path))
            {
                return;
            }
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:403]:
```suggestion
        protected override void BeginProcessing()
```
Please fix below too.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:533]:
```suggestion
        [ValidateNotNullOrEmpty]
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:917]:
```suggestion
        internal StringBuilder _stringBuilder;
```
Perhaps we could use more informative name?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:942]: Duplicate code?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (Shriram0908) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:472]: @iSazonov 
If I change from 
```
            if (Depth == 0)
            {
                _serializer = new Serializer(_xmlWriter);
            }
            else
            {
                _serializer = new Serializer(_xmlWriter, Depth, true);
            }
```
to
```
             _serializer = new Serializer(_xmlWriter, Depth, true);
```
I am getting the below error
```
[-] ConvertFrom-CliXml directive should work 15ms
      PSArgumentException: Value of depth parameter must be greater than or equal to 1.
      at <ScriptBlock>, F:\Source\PowerShell\test\powershell\Modules\Microsoft.PowerShell.Utility\XMLCommand.Tests.ps1: line 164
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (Shriram0908) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:367]: @iSazonov Paging is implement in the import function of ImportXmlHelper class. I am using the same helper used by Import-Clixml cmdlet

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (Shriram0908) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:942]: @iSazonov , I tired to use the same function but I was not sure how to do it so I have created a overloaded the method. This is to make sure that Import-Clixml and ConvertFrom-Clixml use the same helper function.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (Shriram0908) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: I am not sure what to call it. I contains the string array and is used to create the XmlReader  for Deserializer().
Also can we change it to private this will also fix the CodeFactor issue.

**Comment by Shriram0908** [comment]
@iSazonov I have made the requested changes. Please review and provide your feedback.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:471]: It is tricky - SerializationContext() uses 2 as default but cmdlet - zero.
We could fix this in another PR.

Closed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: Please convert to private if possible.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:935]: It is static. So it could be moved to a temp class or a base class.

Comment (Shriram0908) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:935]: class ImportXmlHelper does't not derive from any other class.
I tried to convert sting into memory stream so that we dont need this function itself. Let me know if we can implementation this

```        
internal ImportXmlHelper(PSCmdlet cmdlet, string cliObject)
        {
            _cmdlet = cmdlet;
            _memoryStream = new MemoryStream();
            StreamWriter memeoryWriter = new StreamWriter(_memoryStream);
            memeoryWriter.Write(cliObject);
            memeoryWriter.Flush();
            _memoryStream.Position = 0;
            CreateStream();
        }
```

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:935]: It depends on whether we want to process "#< CLIXML".

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:163]: What do you want test here?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:202]: Indentations

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:233]:
```suggestion
		$content = [PSCustomObject]@{ foo = 1; bar = @{ hello = "world" }} | ConvertTo-Clixml
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:247]:
```suggestion
		$content = @{ a = 1; b = 2; c = 3; d = 4 } | ConvertTo-Clixml
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:263]:
```suggestion
        It "Should import PSCredential" {
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (Shriram0908) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:163]: I copied the same set of tests that was present for ```Import-clixml``` and ```Export-clixml``` and made changes. I didn't write any test. This test seems to be like a basic functionality test.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (Shriram0908) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: Comands like below produce output with "#< CLIXML"
```
Start-Process -FilePath $PSHOME\pwsh -RedirectStandardOutput "temp.xml" -Args "-noprofile -nologo -outputformat xml -command get-command import-clixml" -Wait
```
I have not sure where else this behavior exists.
image

**Comment by vexx32** [comment]
I have one main question regarding the code here (there are minor things, but I think the main thing should be addressed first) -- if I'm not mistaken, a fair amount of parameters and the serialization methods are pretty similar to the existing Export/Import-Csv cmdlets. Have we looked at whether some of the existing code (and/or the new code) can be moved into a base class that both cmdlets can implement to reduce code duplication here? Or are we expecting to do that after this PR?

**Comment by Shriram0908** [comment]
@vexx32, I tried not make too many changes to existing code. I made modifications only to ```ImportXmlHelper``` class (which is used by ```Import-Clixml```) to work with ```ConvertFrom-Clixml```. 

I am following this approach to reduce the PR size. we can have two separate PR or one single PR. I am okay with it either way.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:598]: It's unusual for a `Dispose()` method to throw, but iirc it can happen; I'd prefer to see this with a try/finally so that `_disposed` is always set correctly.

**Review by vexx32 — COMMENTED** [review]
Looks good to me, apart from a couple smaller comments. 🙂

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:598]: It seems `_importXmlHelper.Dispose()` can not throw.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (Shriram0908) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:530]: @vexx32 I changed to ```PSObject``` and tried it code execution seems to follow the same path. No difference either way. Maybe using ```PSObject``` can improve readability.
I used ```PSObject[]``` seeing ```ConvertFrom-Csv```.

Comment (vexx32) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:530]: Yeah @iSazonov I'm sure a lot of commands still do. I'm not sure if it's the right way to go though.

@Shriram0908 if it works both ways... it's probably fine to leave as is. I would prefer to have someone on the PS team weigh in first though.

Comment (Shriram0908) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:530]: @vexx32 Sure okay. let me know anything else needs to be done.

**Review by vexx32 — APPROVED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:552]:
```suggestion
            _stringBuilder.AppendLine(InputObject[0].ToString());
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:167]:
```suggestion
        $content = Get-Command export* -Type Cmdlet | Select-Object -First 3 | ConvertTo-Clixml
        $results = ConvertFrom-CliXml $content
        $results.Count | Should -Be 3
        $results[0].PSTypeNames[0] | Should -BeExactly "Deserialized.System.Management.Automation.CmdletInfo"
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:177]:
```suggestion
    It "ConvertFrom-CliXml with Rehydration should work" {
        $property1 = 256
        $property2 = "abcdef"
        $isHiddenTestType = [IsHiddenTestType]::New($property1,$property2)
        $content = $isHiddenTestType | ConvertTo-Clixml
        $results = ConvertFrom-CliXml $content
        $results.Property1 | Should -Be $property1
        $results.Property2 | Should -BeExactly $property2
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:201]:
```suggestion
    It "ConvertFrom-CliXml StopProcessing should succeed" {
        $content = 1,2,3 | ConvertTo-Clixml
        $ps = [PowerShell]::Create()
        $ps.AddCommand("Get-Process")
        $ps.AddCommand("ConvertFrom-CliXml")
        $ps.AddParameter("InputObject", $content)
        $ps.BeginInvoke()
        $ps.Stop()
        $ps.InvocationStateInfo.State | Should -BeExactly "Stopped"
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:251]:
```suggestion
	It "ConvertFrom-CliXml -First and -Skip work together for collections" {
            $content = @{ a = 1; b = 2; c = 3; d = 4 } | ConvertTo-Clixml
            # order not guaranteed, even with [ordered] so we have to be smart here and compare against the full result
            $out1 = ConvertFrom-CliXml -InputObject $content	# this results in a hashtable
            $out2 = ConvertFrom-CliXml -InputObject $content -First 2 -Skip 1	# this results in a dictionary entry
            $out2.Count | Should -Be 2
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:207]: Indentation.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:233]: Indentation.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:241]: Indentation.

Comment (Shriram0908) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:241]: @iSazonov I just realized that the indentation issues is due to mix of tab and space. Which one should I use space or tab or is it fine as long as I use it consistently ?

Comment (vexx32) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:241]: Spaces is generally the norm in the repo. The only exceptions I usually see are really old files.

**Review by iSazonov — APPROVED** [review]

**Comment by Shriram0908** [comment]
@iSazonov  @vexx32 
What is the next step. Shall I file a new document request.

**Comment by vexx32** [comment]
Yeah, please file a documentation issue in the docs repo and put the link in the PR description. 😊

**Comment by Shriram0908** [comment]
@iSazonov @vexx32 , Please let me know what to do next.

**Comment by iSazonov** [comment]
Waiting MSFT team approve.

**Comment by kvprasoon** [comment]
@iSazonov  Hope `Review - Needed` label is required...

**Comment by bravo-kernel** [comment]
@Shriram0908 thank you for this one, hope it gets merged ❤️

**Comment by Shriram0908** [comment]
@iSazonov @vexx32 @TravisEz13 @SteveL-MSFT 

Its been quite some time. Please let me know if I could help.

**Comment by iSazonov** [comment]
I can not merge without additional reviews.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1
Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:46]: Can we add a secure string?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:987]: review for security issues.

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (TravisEz13) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:987]: adding wg-security for security review.

**Comment by PaulHigin** [comment]
@wg-security
We cannot use the .NET serializer/deserializer as they expose injection issues for PowerShell.  Instead, you should be using the PowerShell serializer APIs.  These APIs are used by Export/Import-CliXml cmdlets and are safe to use.

System.Management.Automation.PSSerializer.Deserialize()
System.Management.Automation.PSSerializer.Serialize()

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:370]:
```suggestion
    public sealed class ConvertToClixmlCommand : PSCmdlet, IDisposable
```

**Comment by Shriram0908** [comment]
How to re-open this PR ?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs
Comment (Shriram0908) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @vexx32 I have made the correction. Thank you for pointing it out.

**Comment by Shriram0908** [comment]
@iSazonov  Thank you for opening the PR. Seems like Link in internals.md and ADOPTERS.md are failing. Let me know if there is something that I can do.

**Comment by iSazonov** [comment]
@Shriram0908 I think you need to rebase.

**Comment by Shriram0908** [comment]
@iSazonov Once I rebased the pushed to repo, It worked. Thanks.

**Comment by SteveL-MSFT** [comment]
It appears that @TravisEz13's feedback hasn't been addressed yet

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
Please address the comments from https://github.com/PowerShell/PowerShell/pull/12845#issuecomment-891349031.

**Pull request closed without merging**
